File: src/kite/scrapers/indian_kanoon.nim
import std/[re, strutils, times]
import chronicles
from ../../tja/base_scraper import BaseScraper, initBaseScraper, makeRequest
from ../../tja/data_models import CaseData, initCaseData
from ../../tja/helpers import sanitizeText, normalizeCourtName


type
  IndianKanoonScraper* = ref object of BaseScraper

method baseUrl*(self: IndianKanoonScraper): string = "https://indiankanoon.org"
method jurisdiction*(self: IndianKanoonScraper): string = "India"

proc newIndianKanoonScraper*(): IndianKanoonScraper = IndianKanoonScraper(initBaseScraper())

proc parseResult(self: IndianKanoonScraper; block: string): CaseData =
  var c = initCaseData()
  let titleRe = re"<a[^>]*class=\"result_title\"[^>]*href=\"([^"]+)\"[^>]*>(.*?)</a>"
  let tm = titleRe.find(block)
  if tm.matchedLen == 0: return c
  let href = block.substr(tm.group(0,1).a, tm.group(0,1).b)
  c.caseName = sanitizeText(block.substr(tm.group(0,2).a, tm.group(0,2).b))
  c.url = if href.startsWith("http"): href else: self.baseUrl() & href
  let idm = re"/doc/(\d+)/".find(c.url)
  if idm.matchedLen > 0: c.caseId = c.url.substr(idm.group(0,1).a, idm.group(0,1).b)
  # meta
  let metaRe = re"<div class=\"result_meta\">([\s\S]*?)</div>"
  let mm = metaRe.find(block)
  if mm.matchedLen > 0:
    let meta = sanitizeText(block.substr(mm.group(0,1).a, mm.group(0,1).b))
    let courtm = re"Court:\s*([^,\n]+)".find(meta, {reIgnoreCase})
    if courtm.matchedLen > 0:
      c.court = normalizeCourtName(meta.substr(courtm.group(0,1).a, courtm.group(0,1).b))
    let dm = re"(\d{1,2}-\d{1,2}-\d{4})".find(meta)
    if dm.matchedLen > 0:
      try: c.date = parse("dd-MM-yyyy", meta.substr(dm.group(0,1).a, dm.group(0,1).b)) except: discard
    for cm in findAll(meta, re"(\d{4})\s+(\d+)\s+(SCC|SCR|AIR)"):
      let y = meta.substr(cm.group(0,1).a, cm.group(0,1).b)
      let n = meta.substr(cm.group(0,2).a, cm.group(0,2).b)
      let rep = meta.substr(cm.group(0,3).a, cm.group(0,3).b)
      c.citations.add "(" & y & ") " & n & " " & rep
  # summary
  let sumRe = re"<div class=\"result_summary\">([\s\S]*?)</div>"
  let sm = sumRe.find(block)
  if sm.matchedLen > 0:
    c.summary = sanitizeText(block.substr(sm.group(0,1).a, sm.group(0,1).b))
  c.jurisdiction = self.jurisdiction()
  return c

proc searchCases*(self: IndianKanoonScraper; query: string; startDate = ""; endDate = ""; court = ""; limit = 100): seq[CaseData] =
  var params: seq[(string,string)] = @[]
  if query.len > 0: params.add ("formInput", query)
  if startDate.len > 0: params.add ("from_date", startDate)
  if endDate.len > 0: params.add ("to_date", endDate)
  if court.len > 0: params.add ("court", court)
  params.add ("limit", $(min(limit, 200)))
  let url = self.baseUrl() & "/search/"
  let resp = self.makeRequest(url, params=params)
  var out: seq[CaseData] = @[]
  for m in findAll(resp.body, re"<div class=\"result\">([\s\S]*?)</div>"):
    if out.len >= limit: break
    out.add self.parseResult(resp.body.substr(m.group(0,1).a, m.group(0,1).b))
  return out

proc getCaseById*(self: IndianKanoonScraper; caseId: string): CaseData =
  if caseId.len == 0: return initCaseData()
  var url: string
  if caseId.startsWith("http"): url = caseId
  elif caseId.allCharsInSet({'0'..'9'}): url = self.baseUrl() & "/doc/" & caseId & "/"
  else:
    let r = self.searchCases(caseId, limit=1); if r.len > 0: return r[0] else: return initCaseData()
  let resp = self.makeRequest(url)
  var c = initCaseData(); c.url = url
  let titleRe = re"<title>(.*?)</title>"; let tm = titleRe.find(resp.body)
  if tm.matchedLen > 0: c.caseName = sanitizeText(resp.body.substr(tm.group(0,1).a, tm.group(0,1).b))
  c.jurisdiction = self.jurisdiction()
  return c

File: internal/scraper/jurisdictions/indiankanoon.go
package jurisdictions

import (
	"context"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/PuerkitoBio/goquery"
	"github.com/gongahkia/kite/internal/scraper"
	"github.com/gongahkia/kite/pkg/errors"
	"github.com/gongahkia/kite/pkg/models"
)

// IndianKanoonScraper implements scraping for Indian Kanoon (India)
type IndianKanoonScraper struct {
	*scraper.BaseScraper
	baseURL string
	client  *http.Client
}

// NewIndianKanoonScraper creates a new Indian Kanoon scraper
func NewIndianKanoonScraper() *IndianKanoonScraper {
	baseURL := "https://indiankanoon.org"
	base := scraper.NewBaseScraper(
		"IndianKanoon",
		"India",
		baseURL,
		10, // Conservative rate limit
	)

	return &IndianKanoonScraper{
		BaseScraper: base,
		baseURL:     baseURL,
		client: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

// SearchCases searches for cases matching the query
func (iks *IndianKanoonScraper) SearchCases(ctx context.Context, query scraper.SearchQuery) ([]*models.Case, error) {
	searchURL, err := iks.buildSearchURL(query)
	if err != nil {
		return nil, errors.ParsingError("failed to build search URL", err)
	}

	allowed, err := iks.BaseScraper.client.CheckRobots(ctx, "/search/")
	if err != nil || !allowed {
		return nil, errors.ErrRobotsDisallowed
	}

	if err := iks.BaseScraper.client.rateLimiter.Wait(ctx); err != nil {
		return nil, errors.RateLimitError("rate limit exceeded")
	}

	req, err := http.NewRequestWithContext(ctx, "GET", searchURL, nil)
	if err != nil {
		return nil, errors.NetworkError("failed to create request", err)
	}

	req.Header.Set("User-Agent", iks.BaseScraper.client.userAgent)

	resp, err := iks.client.Do(req)
	if err != nil {
		return nil, errors.NetworkError("failed to fetch search results", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, errors.NetworkError(fmt.Sprintf("unexpected status code: %d", resp.StatusCode), nil)
	}

	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return nil, errors.ParsingError("failed to parse HTML", err)
	}

	cases := make([]*models.Case, 0)
	doc.Find("div.result").Each(func(i int, s *goquery.Selection) {
		if query.Limit > 0 && len(cases) >= query.Limit {
			return
		}
		caseData := iks.extractCaseFromSearchResult(s)
		if caseData != nil {
			cases = append(cases, caseData)
		}
	})

	return cases, nil
}

// GetCaseByID retrieves a specific case by its ID
func (iks *IndianKanoonScraper) GetCaseByID(ctx context.Context, caseID string) (*models.Case, error) {
	caseURL := iks.buildCaseURL(caseID)

	allowed, err := iks.BaseScraper.client.CheckRobots(ctx, "/doc/")
	if err != nil || !allowed {
		return nil, errors.ErrRobotsDisallowed
	}

	if err := iks.BaseScraper.client.rateLimiter.Wait(ctx); err != nil {
		return nil, errors.RateLimitError("rate limit exceeded")
	}

	req, err := http.NewRequestWithContext(ctx, "GET", caseURL, nil)
	if err != nil {
		return nil, errors.NetworkError("failed to create request", err)
	}

	req.Header.Set("User-Agent", iks.BaseScraper.client.userAgent)

	resp, err := iks.client.Do(req)
	if err != nil {
		return nil, errors.NetworkError("failed to fetch case", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotFound {
		return nil, errors.ErrNotFound
	}

	if resp.StatusCode != http.StatusOK {
		return nil, errors.NetworkError(fmt.Sprintf("unexpected status code: %d", resp.StatusCode), nil)
	}

	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return nil, errors.ParsingError("failed to parse HTML", err)
	}

	return iks.extractCaseDetails(doc, caseID, caseURL)
}

// GetCasesByDateRange retrieves cases within a date range
func (iks *IndianKanoonScraper) GetCasesByDateRange(ctx context.Context, startDate, endDate time.Time, limit int) ([]*models.Case, error) {
	query := scraper.SearchQuery{
		StartDate: &startDate,
		EndDate:   &endDate,
		Limit:     limit,
	}
	return iks.SearchCases(ctx, query)
}

// IsAvailable checks if Indian Kanoon is available
func (iks *IndianKanoonScraper) IsAvailable(ctx context.Context) bool {
	req, err := http.NewRequestWithContext(ctx, "HEAD", iks.baseURL, nil)
	if err != nil {
		return false
	}
	resp, err := iks.client.Do(req)
	if err != nil {
		return false
	}
	defer resp.Body.Close()
	return resp.StatusCode == http.StatusOK
}

// buildSearchURL builds the search URL with query parameters
func (iks *IndianKanoonScraper) buildSearchURL(query scraper.SearchQuery) (string, error) {
	params := url.Values{}

	if query.Query != "" {
		params.Set("formInput", query.Query)
	}

	if query.Court != "" {
		params.Set("court", query.Court)
	}

	if query.StartDate != nil {
		params.Set("fromYear", fmt.Sprintf("%d", query.StartDate.Year()))
	}

	if query.EndDate != nil {
		params.Set("toYear", fmt.Sprintf("%d", query.EndDate.Year()))
	}

	searchURL := fmt.Sprintf("%s/search/?%s", iks.baseURL, params.Encode())
	return searchURL, nil
}

// buildCaseURL builds a case URL from a case ID
func (iks *IndianKanoonScraper) buildCaseURL(caseID string) string {
	caseID = strings.TrimSpace(caseID)
	if strings.HasPrefix(caseID, "http") {
		return caseID
	}
	// Indian Kanoon uses numeric doc IDs
	return fmt.Sprintf("%s/doc/%s/", iks.baseURL, caseID)
}

// extractCaseFromSearchResult extracts case data from a search result item
func (iks *IndianKanoonScraper) extractCaseFromSearchResult(s *goquery.Selection) *models.Case {
	c := models.NewCase()

	titleLink := s.Find("div.result_title a").First()
	caseName := titleLink.Text()
	c.CaseName = strings.TrimSpace(caseName)

	caseURL, exists := titleLink.Attr("href")
	if exists {
		if !strings.HasPrefix(caseURL, "http") {
			caseURL = iks.baseURL + caseURL
		}
		c.URL = caseURL

		// Extract ID from URL: /doc/123456/ -> 123456
		if strings.Contains(caseURL, "/doc/") {
			parts := strings.Split(caseURL, "/doc/")
			if len(parts) >= 2 {
				id := strings.Trim(parts[1], "/")
				c.ID = id
			}
		}
	}

	// Extract court and date from metadata
	metadata := s.Find("div.doc_cite").Text()
	if metadata != "" {
		// Try to extract court
		if strings.Contains(metadata, "Supreme Court") {
			c.Court = "Supreme Court of India"
		} else if strings.Contains(metadata, "High Court") {
			// Extract specific high court
			c.Court = strings.TrimSpace(metadata)
		}
	}

	// Extract snippet
	snippet := s.Find("div.result_highlight").Text()
	c.Summary = strings.TrimSpace(snippet)

	c.Jurisdiction = "India"
	c.SourceDatabase = "IndianKanoon"
	c.ScrapedAt = time.Now()
	c.LastUpdated = time.Now()
	c.Language = "en"
	c.Status = models.CaseStatusActive

	return c
}

// extractCaseDetails extracts detailed case information from a case page
func (iks *IndianKanoonScraper) extractCaseDetails(doc *goquery.Document, caseID, caseURL string) (*models.Case, error) {
	c := models.NewCase()
	c.ID = caseID
	c.URL = caseURL

	// Extract case name from title
	caseName := doc.Find("h1.doc_heading").First().Text()
	if caseName == "" {
		caseName = doc.Find("title").First().Text()
	}
	c.CaseName = strings.TrimSpace(caseName)

	// Extract citation
	citation := doc.Find("div.doc_cite").First().Text()
	c.CaseNumber = strings.TrimSpace(citation)

	// Extract court from citation
	if strings.Contains(citation, "Supreme Court") {
		c.Court = "Supreme Court of India"
	} else if strings.Contains(citation, "Delhi High Court") {
		c.Court = "Delhi High Court"
	} else if strings.Contains(citation, "Bombay High Court") {
		c.Court = "Bombay High Court"
	} else if strings.Contains(citation, "High Court") {
		c.Court = strings.TrimSpace(citation)
	}

	// Extract date
	doc.Find("p, div").Each(func(i int, s *goquery.Selection) {
		text := s.Text()
		if strings.Contains(text, "Decided On:") || strings.Contains(text, "Date:") {
			dateStr := text
			dateStr = strings.ReplaceAll(dateStr, "Decided On:", "")
			dateStr = strings.ReplaceAll(dateStr, "Date:", "")
			dateStr = strings.TrimSpace(dateStr)

			formats := []string{
				"02.01.2006",
				"02-01-2006",
				"2006-01-02",
				"2 January 2006",
			}
			for _, format := range formats {
				if date, err := time.Parse(format, dateStr); err == nil {
					c.DecisionDate = &date
					return
				}
			}
		}
	})

	// Extract judges
	doc.Find("p").Each(func(i int, s *goquery.Selection) {
		text := s.Text()
		if strings.Contains(text, "Bench:") || strings.Contains(text, "Judge:") {
			parts := strings.Split(text, ":")
			if len(parts) > 1 {
				judges := strings.FieldsFunc(parts[1], func(r rune) bool {
					return r == ',' || r == '&'
				})
				for _, judge := range judges {
					judge = strings.TrimSpace(judge)
					if judge != "" {
						c.Judges = append(c.Judges, judge)
					}
				}
			}
		}
	})

	// Extract full judgment text
	fullText := doc.Find("div.judgments").Text()
	if fullText == "" {
		fullText = doc.Find("div.doc_content").Text()
	}
	c.FullText = strings.TrimSpace(fullText)

	c.Jurisdiction = "India"
	c.SourceDatabase = "IndianKanoon"
	c.ScrapedAt = time.Now()
	c.LastUpdated = time.Now()
	c.Language = "en"
	c.Status = models.CaseStatusActive

	return c, nil
}
